### PR TITLE
Redis Cluster support

### LINKED ISSUE #21448: Clustered ownCloud with clustered Redis in HA environment
Hello,

I have an important feature request:

While ownCloud's documentation strongly recommends Redis in a distributed environment, Redis is downright unusable in a HA environment as ownCloud doesn't support a clustered Redis configuration. And that again prevents from using transactional file locking.

So ownCloud is expected to support one of the following:

(1) Support setting multiple Redis masters in config.php (similarly to setting multiple memcached_servers), and try the next one in case the first one fails and so on. This way multiple independent Redis clusters could run, with one being actively used and the others being backups.

(2) Support Redis Sentinel. Redis Sentinel is working in the background and when the current master fails, it propagates a Redis slave to master. In this option, ownCloud calls the Redis Sentinel API to ask for the address of the master.

Thanks a lot.

### PR DESCRIPTION
**Pull request by RobinMcCorkell**
This adds the ability to specify a Redis Cluster, rather than a single Redis server, with the new `redis.cluster` config.php key. There are no unit tests for this since we would need a way to run Docker containers for multiple Redis servers, with the whole kill-a-master-everything-should-work thing.

Note that this does NOT bring support for Redis Sentinel-based configurations, or manual master-slave configurations, that is a different architecture.

Also, there seems to be no way to do authentication or DB index selection with a cluster, for those features you'll want the single server mode

Fixes #21448 

cc @pako81 @cdamken @MorrisJobke @MTRichards

### CODE REVIEW

**Comment by PVince81** [comment]
CC @DeepDiver1975 @icewind1991

File: lib/private/memcache/redis.php [hunk 1]
```diff
@@ -37,32 +37,34 @@ class Redis extends Cache implements IMemcacheTTL {
...
-			} else {
-				$timeout = 0.0; // unlimited
-			}
+			$systemConfig = \OC::$server->getSystemConfig();
 
-			self::$cache->connect($host, $port, $timeout);
-			if(isset($config['password']) && $config['password'] !== '') {
-				self::$cache->auth($config['password']);
-			}
+			if ($config = $systemConfig->getValue('redis.cluster', [])) {
```
Comment (icewind1991) [inline lib/private/memcache/redis.php]: would it makes sense to subclass here instead?

Comment (Xenopathic) [inline lib/private/memcache/redis.php]: Hmm, actually, that's probably a really good idea.

**Comment by karlitschek** [comment]
looks good. :+1:  @DeepDiver1975 ?

**Comment by icewind1991** [comment]
Code looks good, would prefer to have some additional testing before merging though

**Comment by Xenopathic** [comment]
:warning: Currently, phpredis does not have a released version that includes clustering support. Their primary development git branch contains the necessary code, but there's no telling when they decide to release it. It seems that some admins compile their own versions of phpredis, so they get the nice new features... cc @Ethendrel

Not sure how we want to take this? Do enterprise customers want support now, even if that means compiling a development version of phpredis?

**Comment by Ethendrel** [comment]
As an enterprise edition customer I would like to see it implemented rather soon, as transactional file locking is apparently strongly recommended but requires redis. This feature would be needed to maintain the high availability in our clustered environment while using the preferred file locking mechanism.
fyi.. the feature has been running without problems now in our setup for the last 2 days.

**Comment by Xenopathic** [comment]
@Ethendrel Are you sure? That would be surprising to me, since 2.2.7 does not include the cluster code according to GitHub. Could it be that you've got a git-compiled version, but the version code in the extension is still 2.2.7 (they only update the version number at release)?

**Comment by Xenopathic** [comment]
Closed in favour of #24033

**Comment by felixboehm** [comment]
Is there cluster support in future? #24033 was put in backlog, so this might be of interest again.

**Comment by PVince81** [comment]
@DeepDiver1975 since predis support was deferred to future, we should reopen this and revive @Xenopathic's PR for cluster support

**Comment by MorrisJobke** [comment]
> @DeepDiver1975 since predis support was deferred to future, we should reopen this and revive @Xenopathic's PR for cluster support

Implement a completely new interface to just throw it away in the next release (or the one after that) 😕

**Comment by DeepDiver1975** [comment]
> Implement a completely new interface to just throw it away in the next release (or the one after that) :confused:

Agreed - I honestly don't see this in the scope of 9.1 - this needs some serious thinking which usually has to be done before a release.

@karlitschek I'd like to put this to 9.2 and have a discussion first on where we want to go.
@cmonteroluque @MTRichards FYI

**Comment by MTRichards** [comment]
@DeepDiver1975 that sounds like a GREAT idea. This has potentially a serious set of implications...

**Comment by DeepDiver1975** [comment]
I'm closing this for now - we can restore the branch any time if necessary

**Comment by DeepDiver1975** [comment]
rebasing is pointless -> close and restart

**Pull request closed without merging**